File: src/engine/SkipListNodes.go
package engine

type nodes struct {
	nodes []*skipListNode
}

func (nodes *nodes) add(n *skipListNode) {
	nodes.nodes = append(nodes.nodes, n)
}

func (nodes *nodes) pop() *skipListNode {
	n := nodes.nodes[len(nodes.nodes)-1]
	nodes.nodes = nodes.nodes[0 : len(nodes.nodes)-1]
	return n
}

func (nodes *nodes) isEmpty() bool {
	return len(nodes.nodes) == 0
}

File: src/engine/SkipListNodes.cpp
#include "SkipListNodes.h"

void SkipListNodes::add(SkipListNode* node) {
    nodes.push_back(node);
}

void SkipListNodes::addAll(std::vector<SkipListNode*> const &vector) {
    nodes.insert(nodes.end(), vector.begin(), vector.end());
}

bool SkipListNodes::isEmpty() {
    return nodes.size() == 0;
}

SkipListNode* SkipListNodes::pop() {
    SkipListNode* node = nodes.back();
    nodes.pop_back();
    return node;
}